lib/blocks.php

```diff
@@ -433,4 +433,72 @@ function gutenberg_legacy_wp_block_post_meta( $value, $object_id, $meta_key, $si
 
 	return $value;
 }
+
 add_filter( 'default_post_metadata', 'gutenberg_legacy_wp_block_post_meta', 10, 4 );
+
+
+/**
+ * Check if the lightbox should be rendered for a given block.
+ *
+ * This function is INTENTIONALLY left out of core as it only provides
+ * backwards compatibility for the legacy lightbox syntax that was only
+ * introduced in Gutenberg. The legacy syntax was using the `behaviors` key in
+ * the block attrbutes and the `theme.json` file.
+ *
+ * @param array $block The block to check.
+ * @return array The block with the lightboxEnabled flag set if the lightbox should be rendered.
+ */
+function gutenberg_should_render_lightbox( $block ) {
+
+	if ( 'core/image' !== $block['blockName'] ) {
+		return $block;
+	}
+
+	// Get the lightbox setting from the block attributes.
+	if ( isset( $block['attrs']['lightbox'] ) ) {
+		$lightbox_settings = $block['attrs']['lightbox'];
+		// If not present, check legacy syntax.
+	} elseif ( isset( $block['attrs']['behaviors']['lightbox'] ) ) {
+		$lightbox_settings = $block['attrs']['behaviors']['lightbox'];
+	}
+
+	// If not present in block attributes, check global settings.
+	if ( ! isset( $lightbox_settings ) ) {
+		$lightbox_settings = gutenberg_get_global_settings( array( 'lightbox' ), array( 'block_name' => 'core/image' ) );
+
+		// If not present in global settings, check the top-level global settings.
+		if ( ! isset( $lightbox_settings['enabled'] ) ) {
+			$lightbox_settings = gutenberg_get_global_settings( array( 'lightbox' ) );
+		}
+	}
+
+	// If not present in global settings, check legacy syntax.
+	if ( ! isset( $lightbox_settings ) || is_null( $lightbox_settings['enabled'] ) ) {
+		$theme_data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_data();
+		if ( isset( $theme_data['behaviors']['blocks'][ $block['blockName'] ]['lightbox'] ) ) {
+			$lightbox_settings = $theme_data['behaviors']['blocks'][ $block['blockName'] ]['lightbox'];
+		}
+	}
+
+	$link_destination = isset( $block['attrs']['linkDestination'] ) ? $block['attrs']['linkDestination'] : 'none';
+
+	// If the lightbox is enabled, the image is not linked, flag the lightbox to be rendered.
+	if ( isset( $lightbox_settings ) &&
+		true === $lightbox_settings['enabled'] &&
+		'none' === $link_destination
+	) {
+		$block['lightboxEnabled'] = true;
+
+		// The legacy syntax allows setting the animation type.
+		if ( isset( $lightbox_settings['animation'] ) ) {
+			$block['lightboxAnimation'] = $lightbox_settings['animation'];
+		}
+	}
+
+	return $block;
+}
+
+// Run with a priority of 15 to ensure it runs after
+// `block_core_image_should_render_lightbox` core filter (that has a
+// priority of 10).
+add_filter( 'render_block_data', 'gutenberg_should_render_lightbox', 15, 1 );
```

Comment:
Should we move this to the block supports folder, maybe a dedicated lightbox.php file?

Reply:
Does it make sense there? It's not enabled via the `supports` syntax, and is inherent to all images.

I don't have strong opinions either way and am new to working with the blocks in this way, so am happy to go with what's recommended.

Reply:
Oh, so this function is only about the legacy behaviors thing. Sorry I misread, I think it should probably be removed. Why did we decide to keep the legacy syntax working if we never shipped it into Core?

Reply:
It was recommended to us to keep supporting the legacy syntax by @gziolo, as that setting is already widely available in the Gutenberg plugin and many users may have already used it to enable the lightbox.

The reasoning is that it would be a bad UX to drop support for the old syntax and change the behavior of many users' sites, which could cause a drop in confidence in the Gutenberg plugin.

Is there anything I'm missing @gziolo?

Reply:
Ok, that works for me but please let's make sure to remove that before we backport into Core (or not backport this part to Core) to avoid having to deal with this forever.

I think the practice for Gutenberg only APIs have been to keep them around for two Gutenberg releases, deprecate them (a message shown in the console) and then remove the API after two Gutenberg releases. That said, we also did remove some APIs right away when we knew that folks were not using them or that the impact was limited.

Reply:
As an admin, I would be confused if the lightbox I enabled globally for the site in the editor is no longer present on the website, but who knows if people even discovered this currently available setting when using the Gutenberg plugin on production. One solution is to communicate it clearly in the Gutenberg plugin release post and ensure folks are aware of it and they know how to re-enable it. The same applies to the setting integrated with individual Image blocks that in theory might be stored today in the database as:

```html
<!-- wp:image {"id":28,"sizeSlug":"large","linkDestination":"none","behaviors":{"lightbox":{"enabled":true,"animation":"fade"}}} -->
```

This is the only thing we might consider to migrate for users. However, I will leave it to @youknowriad to pick the final approach as he has way better experience with dealing with this type of situations with the site editor.

Reply:
I'd say let's remove the backward-compatibility layer and add a deprecation. 

@ramonjd Also confirmed to us the 2-versions-back rule and [this](https://github.com/WordPress/gutenberg/pull/47475#issuecomment-1409560457) depreciation policy.

I think that the right way to do it is to use the https://github.com/WordPress/gutenberg/tree/update/lightbox-settings-panels/packages/deprecated package and add a deprecation in the editor only, correct?

Reply:
I've just realized that it's impossible to log a deprecation warning for the `behaviors` syntax without re-adding half of https://github.com/WordPress/gutenberg/pull/53851. 

That's because `behaviors` was a new top-level key in `theme.json` and we removed it in that PR. We can still log the deprecation if there is a `behaviors` attribute in the Image block.

Reply:
> Also confirmed to us the 2-versions-back rule and https://github.com/WordPress/gutenberg/pull/47475#issuecomment-1409560457 deprecation policy.

👍🏻 

Just to confirm, I've been told this is the working rule, but I'm not sure it's "official" 😅 Rather, "fluid" depending on the feature it appears. In this case, I'd wager it might apply given that it's only a couple of plugin versions old and not in Core (?)

Reply:
Right, thanks Ramon, that makes sense 🙂

We're going with the most "generous" option anyways! We'll keep the backwards compat for 2 releases and log a deprecation if a `behaviors` attribute is present in a block. We can't detect if a top-level `behaviors` key is used in the `theme.json` so can't log a deprecation in that case, unfortunately.